### PR TITLE
mark functions and vars as static

### PR DESCRIPTION
**Pull request by N-R-K**
the goal here to mark functions and variables not used outside the
translation unit as static. main reason for this is cleanliness. however
as a side-effect this can help compilers optimize better as it now has
guarantee that a certain function won't be called outside of that
translation unit.

### CODE REVIEW

**Comment by N-R-K** [comment]
> With #76 those functions/vars could actually be used and this change would force only those from a single source file to be used at a time.

You can still use them just fine under the right `ifdef`. And you had to `extern` them anyways, so making a one liner wrapper function isn't exactly that much more work.

**Comment by TAAPArthur** [comment]
> You can still use them just fine under the right `ifdef`. And you had to `extern` them anyways, so making a one liner wrapper function isn't exactly that much more work.

No you couldn't access vars defined in 2 separate files in config.h if they were both static. Only one would be in the given translation unit at a time. It works fine if you never want to mix vars from different files but don't know why we'd want that restriction.

For example you couldn't have a function that accessed both `win_draw_bar` function and `resized` with this change since they both will always be defined in distinct translation units.

**Comment by N-R-K** [comment]
> you couldn't have a function that accessed both `win_draw_bar` function and `resized` with this change

You can through a wrapper function. I am not seeing this as any different than having to `extern` declares. Both the wrapper function and extern declare are going to be one line at best.

**Comment by TAAPArthur** [comment]
Maybe you know something I don't
I have checked out your branch and added
```
diff --git a/config.def.h b/config.def.h
index 75c5eed..96ad392 100644
--- a/config.def.h
+++ b/config.def.h
@@ -77,8 +77,16 @@ static const int ignore_mask = Mod2Mask | LockMask;
 /* abort the keyhandler */
 static const KeySym keyhandler_abort = XK_Escape;

+
+void win_draw_bar();
+bool foo() {
+    extern opt_t _options;
+    win_draw_bar(NULL);
+    return _options.hide_bar;
+}
 /* keyboard mappings for image and thumbnail mode: */
 static const keymap_t keys[] = {
+       { 0,            XK_Q,             {foo},               None },
        /* modifiers    key               function              argument */
        { 0,            XK_q,             g_quit,               None },
        { 0,            XK_Return,        g_switch_mode,        None },

```
which is a stupid function that tries to access things declared in 2 different files. My claim is that there is nowhere in config.h you could place foo such that you can compile nsxiv. You can modify config.h anyway you like and if you can find someplace where foo is defined as stated above and bound to some key, I'll retract my argument.

**Comment by N-R-K** [comment]
>  My claim is that there is nowhere in config.h you could place foo such that you can compile nsxiv. You can modify config.h anyway you like and if you can find someplace where foo is defined as stated above and bound to some key, I'll retract my argument.

This compiles fine for me.

```diff
 #ifdef _WINDOW_CONFIG
 
+static void win_draw_bar(win_t *);
+void myWinDraw(void) { win_draw_bar(NULL); }
+bool getHBar(void) { return options->hide_bar; }
+
 /* default window dimensions (overwritten via -g option): */
 static const int WIN_WIDTH  = 800;
 static const int WIN_HEIGHT = 600;
@@ -71,6 +75,12 @@ static const int THUMB_SIZE = 3;
 #endif
 #ifdef _MAPPINGS_CONFIG
 
+bool fun(arg_t _)
+{
+	myWinDraw();
+	return getHBar();
+}
+
 /* following modifiers (NumLock | CapsLock) will be ignored when processing keybindings */
 static const int ignore_mask = Mod2Mask | LockMask;
 
@@ -80,6 +90,7 @@ static const KeySym keyhandler_abort = XK_Escape;
 /* keyboard mappings for image and thumbnail mode: */
 static const keymap_t keys[] = {
 	/* modifiers    key               function              argument */
+	{ 0,            XK_a,             {fun},                None },
 	{ 0,            XK_q,             g_quit,               None },
```

I *will* admit that I didn't realize that I'd need to declare `win_draw_bar`, however the claim that it's not possible is not valid.

**Comment by N-R-K** [comment]
Let me ask a different question, *exactly* which of the vars/functions do you think users will commonly want to access that got turned into `static` in this PR?

One I can think of is `_options`, however you have a pointer `options` to it. You can cast the const away and access it anyways. I don't believe this is undefined behavior since `_options` itself is not declared as `const` only the pointer is.

**Comment by TAAPArthur** [comment]
Have to declare myWinDraw and getHBar but otherwise you'll example meets the criteria I had laid out and I retract my criticism. 

>One I can think of is _options, however you have a pointer options to it. You can cast the const away and access it anyways. I don't believe this is undefined behavior since _options itself is not declared as const, only the pointer is.

You lost me at cast the const away. 

I get what you mean by making the code cleanlier. Marking the helper functions as static seems positive to me. Just not convinced about all the others and the vars. Also trying to track how the non-static vars are used across the code base is more of a problem in my opinion. But to answer your question, `run`, `run_key_handler` seems like bad things to make static. Especially the latter as that could be used in user function. Maybe `check_add_file` too?

And while you made the point that nothing has been lost by default, I think you made it harder. In addition to having to declare the var we additionally have to put it in a wrapper. Think that overrides the cleanliness benefit. Not to mention, do we really expect someone to just know that when editing the config.h?

And it also kills my usecase of using a static lib. It is one thing to deny changes related to that case, but don't think "cleanliness" is a good reason to kill it.

Let me rephrase. I think it is fine to make something static if it really isn't going to be used outside of that file (is a helper method). Think I'm fine with window.c and thumbs.c and don't care about print_version nor inotify.c. Not sure about image.c and still against some/most of the changes in main.c at this moment

**Comment by N-R-K** [comment]
Okay, `run_key_handler` is valid. But `run` ?? Can you show me a usecase where you'd want to call `run` from a custom function?

> do we really expect someone to just know that when editing the config.h

I typically argue in favor of making things noob/amateur friendly, however if someone is writing custom C functions I think we can assume he's at least somewhat familiar with the lang.

> Not sure about image.c and still against some/most of the changes in main.c and _options at this moment

Take your time to go through the patch and list all the changes which you think will be accessed commonly. If the rationale is solid then I won't object to removing static from it, e.g `run_key_handler`.

**Comment by N-R-K** [comment]
On second though, the `run_key_handler` might not be valid. From: https://github.com/nsxiv/nsxiv/issues/120#issuecomment-943402639 We've taken this:
```c
bool send_with_ctrl(arg_t key) {
	extern void run_key_handler(const char *, unsigned);
	run_key_handler(XKeysymToString(key), ControlMask);
	return false;
}
```

And turned it into this

```c
static void run_key_handler(const char *, unsigned);
bool send_with_ctrl(arg_t key) {
	run_key_handler(XKeysymToString(key), ControlMask);
	return false;
}
```

So unlike the other example of `myWinDraw` , in this case no wrapper function would be needed and the custom function LoC stays same (for what that's worth).

**Comment by N-R-K** [comment]
Btw, I didn't comment on the shared lib use-case since I feel like that's going to derail things into yet another pointless argument where neither of us will change our minds. In short, I don't think we should/shouldn't make changes due to 3rd party patches/forks.

This somewhat relates to #148 , I think we should probably resolve that conversation first (and wait for other maintainers to chime in) before getting into this.

**Comment by TAAPArthur** [comment]
>Okay, run_key_handler is valid. But run ?? Can you show me a usecase where you'd want to call run from a custom function?

Maybe not run as is... but if it was broken up into more logic components probably. It isn't though so I'll retract that.

>Take your time to go through the patch and list all the changes which you think will be accessed commonly. 

My focus is more on being accessed reasonably than commonly. Probably biased into to thinking setups are heavily personalized all with unique usecases. But this is a difference in perspective because I'm looking for evidence that something can't be used rather than could be used. Like `win_draw_text(win_t *win, XftDraw *d, const XftColor *color, int x, int y, char *text, int len, int w)` would just be hard to call from config.h compared to `run_key_handler(const char *, unsigned);` which is why it is easier for me to write off the former.

>I typically argue in favor of making things noob/amateur friendly, however if someone is writing custom C functions I think we can assume he's at least somewhat familiar with the lang.

Generally agree, but if I had to choose between jumping through hoops to get something working, manually removing static with sed across the code base or reverting this change, I'd go with one of the latter 2.

You know upon 2nd thought, someone who really hated this could just `define static` in their config.h which also covers fixing config.h. There are 2 cases in main where this is problematic but those could be adjusted. This terrible hack was inspired by what you suggested in the cli PR.

I think I'm at worst neutral to this PR now at least if option wasn't made static. In all honestly think that it is the most useful to modify/access out of everything affected and I don't think there is point in making that static and not the pointer. How I read the goal of this PR was to mark vars/functions as not being accessed outside of the file, but that isn't really true for options because a pointer to it is referenced. But perhaps I should be saying that we should just get rid of the pointer and operate on _options directly; I get why it was done but questioning if it is really helpful.

**Comment by N-R-K** [comment]
> But perhaps I should be saying that we should just get rid of the pointer and operate on _options directly

Since options aren't meant to be changed after initilization in `parse_options` the `const` pointer was used. But since `_options` is not `static` someone can still declare it with `extern` and access it anyways, which is why I added static to it. I'm not sure if we want to operate on _options directly as it goes against the cleanliness goal. Since one can still modify it via `options` pointer (which is valid C afaik) I'm slightly leaned towards keeping it but I won't object removing it.

Which brings me to one question, which of the `_options` member do you think users would want to modify *outside* of options.c ? @TAAPArthur So far most of the things can just be changed via keybinding, e.g `animate`, `hidebar`, `zoom` etc. So I'm struggling to see a usecase where modifying `_options` outside of `options.c` would be needed.

> someone who really hated this could just `define static` in their config.h which also covers fixing config.h.

damn... Looking at the codebase, this would actually work for `image.c` and `window.c` but in `main.c` there's these two:

https://github.com/nsxiv/nsxiv/blob/d0b5005a023da1e436665fc1d795b20246dd9d80/main.c#L292-L295

https://github.com/nsxiv/nsxiv/blob/d0b5005a023da1e436665fc1d795b20246dd9d80/main.c#L631-L636

If someone's feeling really adventurous, then he can probably define

**Comment by N-R-K** [comment]
*For now* moved `_options` inside `parse_options`. For changing default cli args, see: https://github.com/nsxiv/nsxiv/pull/118#issuecomment-955242161

In case you can come up with a valid usecase for wanting to change `_options` value *outside* of options.c then I will revert the change and remove `static` from it.

**Comment by N-R-K** [comment]
Since nothing has been posted, I assume there isn't any use-case for wanting to modify `_options` outside of options.c. So it should be okay to mark it static.

**Comment by XPhyro** [comment]
Although as N-R-K has shown it is still possible to still add your own functions in, the fact that TAAPArthur could not do it at first glance should be an obvious indicator that an arbitrary person that's writing their own functions also may not be able to.

I think it is good practice to mark stuff as `static`, but it may not be the best action in a program that encourages patches & self compilation.

Due to the good aspect of this, I will not actively oppose the addition of `static`s. However, if this PR is to be merged, we definitely need to document it *thoroughly*, including some `config.h` examples.

**Comment by N-R-K** [comment]
> it may not be the best action in a program that encourages patches & self compilation.

The [suckless coding style](https://suckless.org/coding_style) explicitly says to mark functions not used outside of the translation unit as static. So I was assuming people familiar with suckless style code shouldn't have problems.

> However, if this PR is to be merged, we definitely need to document it thoroughly, including some config.h examples.

I assume you've missed this discussion https://github.com/nsxiv/nsxiv/issues/148#issuecomment-960390028 , but in short we agreed not to care about external patches and/or custom user function because

* It adds extra maintenance burden when evaluating a PR.
* Demonstrated [here](https://github.com/nsxiv/nsxiv/pull/118#issuecomment-952489883) , someone can hack into `main()` by doing some `#define` magic. So it's not realistically possible for us to support all arbitrary custom config.

**Comment by XPhyro** [comment]
> So I was assuming people familiar with suckless style code shouldn't have problems.

Fair enough.

> I assume you've missed this discussion

Yes, I've not been able to follow it, and at this point, it's a bit too long for me to get into.

> but in short we agreed not to care about external patches and/or custom user function because

The linked comment states:

> not care about supporting customization in config.h

I see *to care* and *to care about supporting* as different things. Indeed, we cannot support every possible customisation one might have, but we can care about whether they are able to customise or not.

> but in short we agreed not to care about external patches and/or custom user function

If we do not at all care about customisation, then there is no point in developing nsxiv per the suckless style or calling it suckless.

I am on board about not supporting specific patches, but we should care about the overall hackability.

------

As for this PR, I'm not advocating in favour of some arbitrary customisation, I'm requesting that we document our stance. If `config.h` is deemed not patchable, document it. If it is deemed patchable, provide an example. Not having any documentation or example on how to go about patching, or documentation on what parts of nsxiv are patchable will in return waste users' time or add more burden on us via opened issues.

**Comment by N-R-K** [comment]
> If we do not at all care about customisation, then there is no point in developing nsxiv per the suckless style or calling it suckless.

Think you misunderstood, it's not about customization _in general_ but rather just custom functions, since anyone can do any arbitrary things and it's not possible for us to support them all.

> I'm requesting that we document our stance.

Sure, I think that's reasonable. But how do we go about documenting it? In #76 we didn't add any comment or any documentation, but the git commit message contained details and some examples. Do we do the same here and add the examples in the commit message?

**Comment by XPhyro** [comment]
> But how do we go about documenting it? In #76 we didn't add any comment or any documentation, but the git commit message contained details and some examples. Do we do the same here and add the examples in the commit message?

If that was the case, then I think we can merge this in like #76. After that, we can have a dedicated PR for the documentation, perhaps in the manual.

I can probably open a PR this weekend.

**Review by XPhyro — APPROVED** [review]

**Review by eylles — APPROVED** [review]
